src/uucore/src/lib/features/checksum/mod.rs

```diff
@@ -0,0 +1,613 @@
+            Sha512 => "sha512",
+
+            Shake128 => "shake128",
+            Shake256 => "shake256",
+            Blake3 => "blake3",
+        }
+    }
+
+    pub fn is_legacy(self) -> bool {
+        use AlgoKind::*;
+        matches!(self, Sysv | Bsd | Crc | Crc32b)
+    }
+}
+
+/// Holds a length for a SHA2 of SHA3 algorithm kind.
+#[derive(Debug, Clone, Copy)]
+pub enum ShaLength {
+    Len224,
+    Len256,
+    Len384,
+    Len512,
+}
+
+impl ShaLength {
+    pub fn as_usize(self) -> usize {
+        match self {
+            Self::Len224 => 224,
+            Self::Len256 => 256,
+            Self::Len384 => 384,
+            Self::Len512 => 512,
+        }
+    }
+}
+
+impl TryFrom<usize> for ShaLength {
+    type Error = ChecksumError;
+
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        use ShaLength::*;
+        match value {
+            224 => Ok(Len224),
+            256 => Ok(Len256),
+            384 => Ok(Len384),
+            512 => Ok(Len512),
+            _ => Err(ChecksumError::InvalidLengthForSha(value.to_string())),
+        }
+    }
+}
+
+/// Represents an actual determined algorithm.
+#[derive(Debug, Clone, Copy)]
+pub enum SizedAlgoKind {
+    Sysv,
+    Bsd,
+    Crc,
+    Crc32b,
+    Md5,
+    Sm3,
+    Sha1,
+    Blake3,
+    Sha2(ShaLength),
+    Sha3(ShaLength),
+    // Note: we store Blake2b's length as BYTES.
+    Blake2b(Option<usize>),
+    Shake128(usize),
+    Shake256(usize),
+}
+
+impl SizedAlgoKind {
+    pub fn from_unsized(kind: AlgoKind, byte_length: Option<usize>) -> UResult<Self> {
+        use AlgoKind as ak;
+        match (kind, byte_length) {
+            (
+                ak::Sysv
+                | ak::Bsd
+                | ak::Crc
+                | ak::Crc32b
+                | ak::Md5
+                | ak::Sm3
+                | ak::Sha1
+                | ak::Blake3
+                | ak::Sha224
+                | ak::Sha256
+                | ak::Sha384
+                | ak::Sha512,
+                Some(_),
+            ) => Err(ChecksumError::LengthOnlyForBlake2bSha2Sha3.into()),
+
+            (ak::Sysv, _) => Ok(Self::Sysv),
+            (ak::Bsd, _) => Ok(Self::Bsd),
+            (ak::Crc, _) => Ok(Self::Crc),
+            (ak::Crc32b, _) => Ok(Self::Crc32b),
+            (ak::Md5, _) => Ok(Self::Md5),
+            (ak::Sm3, _) => Ok(Self::Sm3),
+            (ak::Sha1, _) => Ok(Self::Sha1),
+            (ak::Blake3, _) => Ok(Self::Blake3),
+
+            (ak::Shake128, Some(l)) => Ok(Self::Shake128(l)),
+            (ak::Shake256, Some(l)) => Ok(Self::Shake256(l)),
+            (ak::Sha2, Some(l)) => Ok(Self::Sha2(ShaLength::try_from(l)?)),
+            (ak::Sha3, Some(l)) => Ok(Self::Sha3(ShaLength::try_from(l)?)),
+            (algo @ (ak::Sha2 | ak::Sha3), None) => {
+                Err(ChecksumError::LengthRequiredForSha(algo.to_lowercase().into()).into())
+            }
+            // [`calculate_blake2b_length`] expects a length in bits but we
```

Comment:
Comment says 'expects a length in bits' but the multiplication by 8 suggests the input is in bytes ?